lock/src/main/java/org/infinispan/lock/api/impl/lock/ClusteredLockImpl.java

```diff
@@ -0,0 +1,359 @@
+
+/**
+ * Implements {@link ClusteredLock} interface.
+ * <p>
+ * This lock implements a non rentrant where the ownership is {@link org.infinispan.lock.api.OwnershipLevel#NODE}.
+ * <p>
+ * <h2>Non reentrant lock, Owner Node</h2> <lu> <li>Originator of the requests is the {@link Address} of the {@link
+ * org.infinispan.manager.EmbeddedCacheManager}</li> <li>When a lock is acquired by a Node, it cannot be re-acquired,
+ * even by the actual node til the lock is released.</li> <li>The lock can be unlocked only by the lock owner, in this
+ * case the node</li> <li>lock method does not expire til the lock is acquired, so this can cause thread starvation in
+ * the actual implementation</li> </lu>
+ * <p>
+ * <h2>Partition handling</h2>
+ *
+ * @author Katia Aresti, karesti@redhat.com
+ * @see <a href="http://infinispan.org/documentation/">Infinispan documentation</a>
+ * @since 9.2
+ */
+public class ClusteredLockImpl implements ClusteredLock {
+   private static final Log log = LogFactory.getLog(ClusteredLockImpl.class, Log.class);
+
+   private final String name;
+   private final ClusteredLockKey lockKey;
+   private final EmbeddedClusteredLockManager clusteredLockManager;
+   private final FunctionalMap.ReadWriteMap<ClusteredLockKey, ClusteredLockValue> readWriteMap;
+   private final Queue<RequestHolder> pendingRequests;
+   private final Object originator;
+
+   public ClusteredLockImpl(String name,
+                            ClusteredLockKey lockKey,
+                            AdvancedCache<ClusteredLockKey, ClusteredLockValue> clusteredLockCache,
+                            EmbeddedClusteredLockManager clusteredLockManager) {
+      this.name = name;
+      this.lockKey = lockKey;
+      this.clusteredLockManager = clusteredLockManager;
+      this.pendingRequests = new ConcurrentLinkedQueue<>();
+      this.readWriteMap = ReadWriteMapImpl.create(FunctionalMapImpl.create(clusteredLockCache));
+      originator = clusteredLockCache.getCacheManager().getAddress();
+      clusteredLockCache.getCacheManager().addListener(new ClusterChangeListener());
+      clusteredLockCache.addListener(new LockReleasedListener(), new ClusteredLockFilter(lockKey));
+   }
+
+   public abstract class RequestHolder<E> {
+      protected final CompletableFuture<E> request;
+      protected final String requestId;
+      protected final Object requestor;
+
+      public RequestHolder(Object requestor, CompletableFuture<E> request) {
+         this.requestId = createRequestId();
+         this.requestor = requestor;
+         this.request = request;
+      }
+
+      public boolean isDone() {
+         return request.isDone();
+      }
+
+      public void handleLockResult(Boolean result, Throwable ex) {
+         if (ex != null) {
+            log.trace("Exception on lock request " + this, ex);
+            request.completeExceptionally(handleException(ex));
+            return;
+         }
+
+         if (request == null) {
```

Comment:
bug: should `result == null`

Reply:
Indeed ! I could not reproduce this case in tests haha thank you pointing this out.